metricplugin/wiretap.go

```diff
@@ -73,9 +75,35 @@ func (wt *BitswapWireTap) MessageReceived(pid peer.ID, msg bsmsg.BitSwapMessage)
 	   	fmt.Printf("Received msg from %s with multiaddrs: %s \n", pid, mas) */
 
 	now := time.Now()
+
+	// Get CIDs of contained blocks.
+	msgBlocks := msg.Blocks()
+	blocks := make([]cid.Cid, 0, len(msgBlocks))
+	for _, block := range msgBlocks {
+		blocks = append(blocks, block.Cid())
+	}
+
+	// Get block presences.
+	msgBlockPresences := msg.BlockPresences()
+	blockPresences := make([]BlockPresence, 0, len(msgBlockPresences))
+	for _, presence := range msgBlockPresences {
+		p := BlockPresence{
+			Cid: presence.Cid,
+		}
+		if presence.Type == pbmsg.Message_Have {
+			p.Type = Have
+		} else {
+			p.Type = DontHave
+		}
+		blockPresences = append(blockPresences, p)
+	}
+
+	// Construct the message to push to subscribers
 	msgToLog := BitswapMessage{
 		WantlistEntries: msg.Wantlist(),
 		FullWantList:    msg.Full(),
+		Blocks:          blocks,
```

Comment:
This should me mutually exclusive, right? So if we push a msg with Blocks & Presences, we don't have a wantlist and vice versa?
If so, wouldn't it be nicer to split this into two separate messages?

Reply:
Good question!
From the [message protobuf spec](https://github.com/ipfs/go-bitswap/blob/master/message/pb/message.proto) this is not mutually exclusive.
There is also an outdated [bitswap spec](https://github.com/ipfs/specs/blob/master/BITSWAP.md), which says:

> A single Bitswap message may contain any of the following content:
>
> 1.   The sender’s wantlist. This wantlist may either be the sender’s complete wantlist or just the changes to the sender’s wantlist that the receiver needs to know.
> 2.   Data blocks. These are meant to be blocks that the receiver has requested (i.e., blocks on that are on the receiver’s wantlist as far as the sender is aware at the time of sending).

So that doesn't even know about block presences yet.

Digging through the Go implementation it seems that
- The [bitswap engine](https://github.com/ipfs/go-bitswap/blob/master/internal/decision/engine.go#L472-L541) sends out messages containing blocks and block presences (any combination of those). The envelope is picked up and sent out [by a task worker](https://github.com/ipfs/go-bitswap/blob/master/workers.go#L46-L62).
- The [per-peer message queues](https://github.com/ipfs/go-bitswap/blob/master/internal/messagequeue/messagequeue.go#L270-L360) are only responsible for wantlist-related things.

So, on the Go side of things a message can either be wantlist updates or a combination of blocks and block presences.

I also looked at [ipfs-rust/libp2p-bitswap](https://github.com/ipfs-rust/libp2p-bitswap/blob/master/src/compat/message.rs): They expect messages to contain all of the above, in one message. They send messages that are strictly mutually exclusive in their contents, i.e., either a block, a (singular) block presence, or a (singular) wantlist entry.

So yeah, I'd love to split them, but I don't think any spec guarantees mutual exclusivity.

Reply:
I suppose splitting it into (Wantlist|Block+Block Presences) would be reasonable though, having them both be their own struct. Then we reference both of those (nullable) and see where we go from there.
That way we at least have the guarantee that the full wantlist flag is only considered valid if there were wantlist entries.

This does not make them mutually exclusive, but might clean up some things. It might turn out, in practice, that they always are exclusive, we'll have to see.

What do you think?

Reply:
Nice deep dive there, thank you! :) Splitting it up into separate structs sounds good, but I wouldn't prioritize it, to be honest, as I'm unsure if the subsequent processing and analysis code really benefits that much from it.